src/pages/tap.tsx

```diff
@@ -146,48 +140,102 @@ export default function Tap() {
-            }
-            await handleLocationTap(tapResponse.location);
-            break;
-          default:
-            throw new Error("Invalid tap response code!");
-        }
+        .then((response) => {
+          if (!response.ok)
+            throw new Error(`HTTP error! status: ${response.status}`);
+          return response.json();
+        })
+        .then(async (data) => {
+          const tapResponse = tapResponseSchema.validateSync(data);
+          switch (tapResponse.code) {
+            case TapResponseCode.CMAC_INVALID:
+              throw new Error("CMAC invalid!");
+            case TapResponseCode.PERSON_NOT_REGISTERED:
+              handlePersonRegistration(cmac);
+              break;
+            case TapResponseCode.LOCATION_NOT_REGISTERED:
+              handleLocationRegistration(cmac);
+              break;
+            case TapResponseCode.VALID_PERSON:
+              if (!tapResponse.person) {
+                throw new Error("Person is null!");
+              }
+              await handlePersonTap(tapResponse.person);
+              break;
+            case TapResponseCode.VALID_LOCATION:
+              if (!tapResponse.location) {
+                throw new Error("Location is null!");
+              }
+              await handleLocationTap(tapResponse.location);
+              break;
+            default:
+              throw new Error("Invalid tap response code!");
+          }
+        })
+        .catch((error) => {
+          console.error(error);
+          toast.error("Error! Please refresh and try again.");
+        });
+    } else {
+      // ----- HANDLE CARD GENERATED SIGNATURE TAP -----
+      if (!location.hash) {
+        toast.error("Unable to process cmac or signature from tap.");
+        router.push("/");
+        return;
+      }
+
+      const urlParams = new URLSearchParams(location.hash.slice(1));
+      const rawLocationSignature = getHaLoArgs(urlParams);
+      if (!rawLocationSignature) {
+        toast.error("Unable to process cmac or signature from tap.");
+        router.push("/");
+        return;
+      }
+      const { signaturePublicKey, signatureMessage, signature } =
+        rawLocationSignature;
+
+      fetch(`/api/tap/sig_card?signaturePublicKey=${signaturePublicKey}`, {
+        method: "GET",
+        headers: {
+          "Content-Type": "application/json",
+        },
       })
-      .catch((error) => {
-        console.error(error);
-        toast.error("Error! Please refresh and try again.");
-      });
+        .then((response) => {
+          if (!response.ok)
+            throw new Error(`HTTP error! status: ${response.status}`);
+          return response.json();
+        })
+        .then(async (data) => {
+          const sigCardTapResponse =
+            sigCardTapResponseSchema.validateSync(data);
+          if (!sigCardTapResponse.registered) {
+            throw new Error("This location card is not registered!");
```

Comment:
Do we have an issue for registering these?

Reply:
#115